packages/toolpad-app/src/toolpad/AppState.tsx

```diff
@@ -294,6 +329,327 @@ export function appStateReducer(state: AppState, action: AppStateAction): AppSta
         hasUnsavedChanges: action.hasUnsavedChanges,
       });
     }
+    case 'SAVE_QUERY_DRAFT': {
+      if (state.currentView.kind === 'page' && state.currentView.view?.kind === 'query') {
+        const queryTabs = state.currentView.queryPanel?.queryTabs || [];
+        const currentTabIndex = state.currentView.queryPanel?.currentTabIndex;
+        if (currentTabIndex !== undefined && queryTabs) {
+          return update(state, {
+            currentView: {
+              ...state.currentView,
+              queryPanel: {
+                ...state.currentView.queryPanel,
+                queryTabs: state.currentView?.queryPanel?.queryTabs?.map((tab, index) => {
```

Comment:
Maybe we can create some utility that makes operations like this shorter and more readable?